src/engine/structures.rs

```diff
@@ -66,117 +125,108 @@ impl<T: HasName + HasUuid> Table<T> {
-                self.uuid_map.insert(last_item.uuid(), index);
-            }
-
-            // Remove the item we want to remove and also the uuid mapping
-            let item = self.items.swap_remove(index);
-            self.uuid_map.remove(&item.uuid());
-
-            // Remove the name again, in case there is only one item.
-            self.name_map.remove(name);
-
-            Some(item)
+    pub fn remove_by_name(&mut self, name: &str) -> Option<(Uuid, T)> {
+        if let Some(uuid) = self.name_to_uuid.remove(name) {
+            self.items.remove(&uuid).map(|(_, item)| (uuid, item))
         } else {
             None
         }
     }
 
     /// Removes the item corresponding to the uuid if there is one.
-    pub fn remove_by_uuid(&mut self, uuid: Uuid) -> Option<T> {
-        if let Some(index) = self.uuid_map.remove(&uuid) {
-            // Insert mappings for the about-to-be swapped element
-            {
-                let last_item = &self.items
-                                     .last()
-                                     .expect("uuid_map is non-empty <-> items is non-empty");
-                self.name_map.insert(last_item.name().into(), index);
-                self.uuid_map.insert(last_item.uuid(), index);
-            }
-
-            // Remove the item we want to remove and also the uuid mapping
-            let item = self.items.swap_remove(index);
-            self.name_map.remove(item.name());
-
-            // Remove the uuid again, in case there is only one item.
-            self.uuid_map.remove(&uuid);
-
-            Some(item)
+    pub fn remove_by_uuid(&mut self, uuid: Uuid) -> Option<(Name, T)> {
+        if let Some((name, item)) = self.items.remove(&uuid) {
+            self.name_to_uuid.remove(&name);
+            Some((name, item))
         } else {
             None
         }
     }
 
     /// Inserts an item for given uuid and name.
-    /// Returns a list of the items displaced, which may be empty if no items
-    /// are displaced, have one entry if the uuid and the name map to the same
-    /// item, and may have two entries if the uuid and the name map to
-    /// different items.
-    pub fn insert(&mut self, item: T) -> Vec<T> {
-        let name_item = self.remove_by_name(item.name());
-        let uuid_item = self.remove_by_uuid(item.uuid());
-
-        let future_last_index = self.items.len();
-        self.name_map
-            .insert(item.name().into(), future_last_index);
-        self.uuid_map.insert(item.uuid(), future_last_index);
-
-        self.items.push(item);
-
-        match (name_item, uuid_item) {
-            (None, None) => vec![],
-            (None, Some(uuid_item)) => vec![uuid_item],
-            (Some(name_item), None) => vec![name_item],
-            (Some(name_item), Some(uuid_item)) => vec![name_item, uuid_item],
+    /// Possibly returns the item displaced.
+    pub fn insert(&mut self, name: Name, uuid: Uuid, item: T) -> Option<(Name, Uuid, T)> {
```

Comment:
I don't see any way that this can be right. If the name and the uuid are both already in the table, and they don't belong to the same item, then in order to insert the new item, and maintain the invariant that each uuid maps to its name and each name maps to its uuid then two items must be ejected.

In this specific implementation, you are ejecting both, but only returning one.

Reply:
ok